src/app/application/services/data-source/DataSource.ts

```diff
@@ -0,0 +1,83 @@
+import CallableEventSource, { isClosed } from './CallableEventSource'
+import Router from './Router'
+import SharedPool from './SharedPool'
+
+export type DataSourceResponse<T> = {data: T | undefined, error: Error | undefined, refresh: () => void}
+export type Source = (params: Record<string, unknown>, source: {close: () => void}) => Promise<unknown>
+export type Sources = Record<string, Source>
+
+const get = (src: string, router: Router<Source>) => {
+  const [path, query] = src.split('?')
+  const queryParams = new URLSearchParams(query)
+  const route = router.match(path)
+  const _source = new CallableEventSource(async function * (this: CallableEventSource) {
+    while (true) {
+      this.readyState = 1
+      // latency
+      // await new Promise(resolve => setTimeout(resolve, 5000))
+      yield route.route({
+        ...{
+          offset: parseInt(queryParams.get('offset') || '0'),
+        },
+        ...route.params,
+      }, this)
+      if (!isClosed(this)) {
+        await new Promise(resolve => setTimeout(resolve, 5000))
```

Comment:
If we aren't closed after a `Source` resolves its promise, then we are polling, at the moment the poll interval is hardcoded to 5s.